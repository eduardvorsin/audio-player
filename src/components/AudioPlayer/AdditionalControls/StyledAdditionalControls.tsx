import styled from "styled-components";

interface StyledAdditionalControlsProps {
	className?: string,
}

export const StyledAdditionalControls = styled.div<StyledAdditionalControlsProps>`
	display: flex;
	align-items: center;
	justify-content: space-between;

	&>*:not(:last-child){
		margin-right:20px;
	}
`;