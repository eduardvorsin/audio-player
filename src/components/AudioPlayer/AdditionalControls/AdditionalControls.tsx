import React from "react";
import { StyledAdditionalControls } from "./StyledAdditionalControls";
import { Button } from "../../UI/Button/Button";
import { PlaybackRateButton } from "../PlaybackRateButton/PlaybackRateButton";
import { ReactComponent as RepeatIcon } from '../../../assets/images/icons/repeat.svg';
import { ReactComponent as NoRepeatIcon } from '../../../assets/images/icons/no-repeat.svg';
import { ReactComponent as DownloadIcon } from '../../../assets/images/icons/download.svg';

interface AdditionalControlsProps {
  downloadLink?: string,
  className?: string,
  playbackRate?: number,
  showDownloadControl?: boolean,
  showPlaybackRateControl?: boolean,
  showLoopControl?: boolean,
  isLooped?: boolean,
  changePlaybackRate?: React.MouseEventHandler<HTMLButtonElement>,
  changeLooping?: React.MouseEventHandler<HTMLButtonElement>,
}

export const AdditionalControls = React.memo<AdditionalControlsProps>(({
  className = '',
  playbackRate = 1,
  downloadLink = '',
  showDownloadControl = false,
  showPlaybackRateControl = false,
  showLoopControl = false,
  isLooped = false,
  changePlaybackRate = () => { },
  changeLooping = () => { },
}) => {

  const repeatButtonIcon = isLooped ? <RepeatIcon /> : <NoRepeatIcon />;

  return (
    <StyledAdditionalControls
      className={className}
      data-testid='additional-controls'
    >
      {showDownloadControl &&
        <Button
          as='a'
          withoutVisibleText
          href={downloadLink}
          startIcon={<DownloadIcon />}
          download
        >
          download track
        </Button>
      }

      {showPlaybackRateControl &&
        <PlaybackRateButton
          aria-label='current playback rate'
          onClick={changePlaybackRate}
        >
          {`${playbackRate}x`}
        </PlaybackRateButton>
      }

      {showLoopControl &&
        <Button
          withoutVisibleText
          startIcon={repeatButtonIcon}
          onClick={changeLooping}
        >
          {isLooped ? 'don\'t repeat the song' : 'repeat song infinite'}
        </Button>
      }
    </StyledAdditionalControls>
  );
});