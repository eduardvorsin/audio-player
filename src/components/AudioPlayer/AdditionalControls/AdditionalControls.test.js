import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { AdditionalControls } from "./AdditionalControls";

describe('AdditionalControls component tests', () => {
  test('renders correctly', () => {
    const mockFn = jest.fn();
    render(<AdditionalControls
      showDownloadControl
      showPlaybackRateControl
      showLoopControl
      changePlaybackRate={mockFn}
      changeLooping={mockFn}
    />);

    expect(screen.getAllByRole('button')).toHaveLength(3);
  });

  test('renders without passed props', () => {
    render(<AdditionalControls />);

    expect(screen.queryByRole('button')).not.toBeInTheDocument();
  });

  test('when the playback rate button is clicked, the mock function is triggered', async () => {
    const user = userEvent.setup();
    const mockFn = jest.fn();
    render(
      <AdditionalControls
        showPlaybackRateControl
        changePlaybackRate={mockFn}
      />
    );

    await user.click(screen.getByRole('button'));

    expect(mockFn).toBeCalledTimes(1);
  });

  test('when the loop button is clicked, the mock function is triggered', async () => {
    const user = userEvent.setup();
    const mockFn = jest.fn();
    render(<AdditionalControls
      showLoopControl
      changeLooping={mockFn}
    />);

    await user.click(screen.getByRole('button'));

    expect(mockFn).toBeCalledTimes(1);
  });


  test('when the isLooped prop is passed, the text in the loop button changes to don\'t repeat the song', () => {
    render(<AdditionalControls
      isLooped
      showLoopControl
    />);

    expect(screen.getByRole('button')).toHaveTextContent('don\'t repeat the song');
  });

  test('playbackRate prop value correctly passed', () => {
    render(<AdditionalControls
      showPlaybackRateControl
      playbackRate={1.5}
    />);

    expect(screen.getByRole('button')).toHaveTextContent('1.5x');
  });

  test('downloadLink prop value correctly passed', () => {
    render(<AdditionalControls
      showDownloadControl
      downloadLink='abc'
    />);

    expect(screen.getByRole('link').href).toMatch(/(abc)+/i);
  });

  test('snapshot with default prop values', () => {
    render(
      <AdditionalControls />
    );

    expect(screen.getByTestId('additional-controls')).toMatchSnapshot();
  });

  test('snapshot with showPlaybackRateControl prop', () => {
    render(
      <AdditionalControls
        showPlaybackRateControl
      />
    );

    expect(screen.getByTestId('additional-controls')).toMatchSnapshot();
  });

  test('snapshot with showPlaybackRateControl & playbackRate props', () => {
    render(
      <AdditionalControls
        showPlaybackRateControl
        playbackRate={1.5}
      />
    );

    expect(screen.getByTestId('additional-controls')).toMatchSnapshot();
  });

  test('snapshot with showLoopControl prop', () => {
    render(
      <AdditionalControls
        showLoopControl
      />
    );

    expect(screen.getByTestId('additional-controls')).toMatchSnapshot();
  });

  test('snapshot showLoopControl & isLooped props', () => {
    render(
      <AdditionalControls
        showLoopControl
        isLooped
      />
    );

    expect(screen.getByTestId('additional-controls')).toMatchSnapshot();
  });

  test('snapshot showDownloadControl & downloadLink props', () => {
    render(
      <AdditionalControls
        showDownloadControl
        downloadLink='test.ogg'
      />
    );

    expect(screen.getByTestId('additional-controls')).toMatchSnapshot();
  });

  test('snapshot with all the props affecting the UI', () => {
    render(
      <AdditionalControls
        downloadLink='test.wav'
        showDownloadControl
        showPlaybackRateControl
        showLoopControl
        className='test'
      />
    );

    expect(screen.getByTestId('additional-controls')).toMatchSnapshot();
  });
});
