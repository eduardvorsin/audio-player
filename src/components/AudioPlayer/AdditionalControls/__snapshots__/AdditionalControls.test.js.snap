// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalControls component tests snapshot showDownloadControl & downloadLink props 1`] = `
<div
  class="sc-bcXHqe fDkhxT"
  data-testid="additional-controls"
>
  <a
    class="sc-gswNZR ivOWdj"
    data-testid="button"
    download=""
    href="test.ogg"
  >
    <svg />
    download track
  </a>
</div>
`;

exports[`AdditionalControls component tests snapshot showLoopControl & isLooped props 1`] = `
<div
  class="sc-bcXHqe fDkhxT"
  data-testid="additional-controls"
>
  <button
    class="sc-gswNZR ivOWdj"
    data-testid="button"
  >
    <svg />
    don't repeat the song
  </button>
</div>
`;

exports[`AdditionalControls component tests snapshot with all the props affecting the UI 1`] = `
<div
  class="sc-bcXHqe fDkhxT"
  data-testid="additional-controls"
>
  <a
    class="sc-gswNZR ivOWdj"
    data-testid="button"
    download=""
    href="test.wav"
  >
    <svg />
    download track
  </a>
  <button
    aria-label="current playback rate"
    class="sc-gswNZR fzhZbU sc-dkrFOg egzGNd"
    data-testid="button"
  >
    1x
  </button>
  <button
    class="sc-gswNZR ivOWdj"
    data-testid="button"
  >
    <svg />
    repeat song infinite
  </button>
</div>
`;

exports[`AdditionalControls component tests snapshot with default prop values 1`] = `
<div
  class="sc-bcXHqe fDkhxT"
  data-testid="additional-controls"
/>
`;

exports[`AdditionalControls component tests snapshot with showLoopControl prop 1`] = `
<div
  class="sc-bcXHqe fDkhxT"
  data-testid="additional-controls"
>
  <button
    class="sc-gswNZR ivOWdj"
    data-testid="button"
  >
    <svg />
    repeat song infinite
  </button>
</div>
`;

exports[`AdditionalControls component tests snapshot with showPlaybackRateControl & playbackRate props 1`] = `
<div
  class="sc-bcXHqe fDkhxT"
  data-testid="additional-controls"
>
  <button
    aria-label="current playback rate"
    class="sc-gswNZR fzhZbU sc-dkrFOg egzGNd"
    data-testid="button"
  >
    1.5x
  </button>
</div>
`;

exports[`AdditionalControls component tests snapshot with showPlaybackRateControl prop 1`] = `
<div
  class="sc-bcXHqe fDkhxT"
  data-testid="additional-controls"
>
  <button
    aria-label="current playback rate"
    class="sc-gswNZR fzhZbU sc-dkrFOg egzGNd"
    data-testid="button"
  >
    1x
  </button>
</div>
`;
